tools/make_patch.c

```diff
@@ -0,0 +1,828 @@
+				putc('{', output);
+				line_pos++;
+				ungetc(c, file);
+				break;
+			}
+
+			// If we have, we store the contents before it ends into buffer
+			buffer_index = 0;
+			while ((c = getc(file)) != EOF) {
+				if (c == '}') {
+					if ((c = getc(file)) == EOF) break;
+					if (c == '}') break;
+				}
+
+				buffer[buffer_index++] = c;
+				buffer = expand_buffer(buffer, buffer_index);
+				if (!buffer) goto error;
+			}
+
+			buffer[buffer_index] = '\0';
+
+			// If we didn't manage to find the symbol,
+			//   we can just end inmediately.
+			if (!symbol) break;
+
+			patch->offset = 0;
+			interpret_command(buffer, symbol, patch, new_rom, orig_rom, output);
+			if (patch->offset) patch++;
+			patches = expand_buffer(patches, (uintptr_t)patch - (uintptr_t)patches);
+			if (!patches) goto error;
+
+			break;
+
+		case '[':
+			// End inmediately if this is not the beginning of the line.
+			if (line_pos) {
+				putc(c, output);
+				line_pos++;
+				break;
+			}
+
+			// Try to read the label
+			putc(c, output);
+			buffer_index = 0;
+
+			while ((c = getc(file)) != EOF) {
+				putc(c, output);
+
+				if (c == ']') {
+					// If we're at the end, we can get the symbol for the label
+					buffer[buffer_index] = '\0';
+
+					// Translate spaces to underscores
+					for (char *p = buffer; *p != '\0'; p++) {
+						if (*p == ' ') *p = '_';
+					}
+
+					// Look for the symbol
+					symbol = find_symbol(symbols, buffer);
+					if (!symbol) {
+						fprintf(stderr, "Error: Cannot find symbol: %s\n",
+						        buffer);
+					}
+
+					// Skip until the next newline
+					while ((c = getc(file)) != EOF) {
+						putc(c, output);
+						if (c == '\n') break;
+					}
+					buffer_index = 0;
+					break;
+				}
+
+				buffer[buffer_index++] = c;
+			}
+
+			break;
+
+		default:
+			putc(c, output);
+			line_pos++;
+		}
+	}
+
+	patch->offset = 0;
+	free_buffer(buffer);
+	return patches;
+
+error:
+	if (patches) free_buffer(patches);
+	if (buffer) free_buffer(buffer);
+	return NULL;
+}
+
+int compare_patch(const void *_patch1, const void *_patch2)
+{
+	const struct patch *patch1 = _patch1;
+	const struct patch *patch2 = _patch2;
+	return patch1->offset - patch2->offset;
+}
+
+int verify_completeness(FILE *orig_rom, FILE *new_rom, struct patch *patches)
+{
+	struct patch *patch = patches;
+	size_t offset = 0;
+	int c, d;
+
+	while (patch->offset != 0) patch++;
+	qsort(patches, patch - patches, sizeof(struct patch), compare_patch);
+
+	rewind(orig_rom);
+	rewind(new_rom);
+	patch = patches;
+	for (;;) {
+		c = getc(orig_rom);
+		d = getc(new_rom);
+		if (c == EOF || d == EOF) break;
+		if (patch->offset && patch->offset == offset) {
+			if (fseek(orig_rom, patch->size, SEEK_CUR) != 0) return -1;
+			if (fseek(new_rom, patch->size, SEEK_CUR) != 0) return -1;
+
+			offset += patch->size + 1;
+			patch++;
+			continue;
+		}
+		offset++;
+		if (c != d && ((offset-1) < 2096608)) { // offset higher than 2096608 is stadium checksum...
```

Comment:
@mid-kid said in his first post:
>Instead of hardcoding the offset of the stadium checksum in verify_completeness, hardcode it in the same location as the other checksums, the start of process_template. Its offset may have to be calculated the same way as stadium.c does, though I've been considering hardcoding the address in there too, instead of expanding it depending on the ROM size like is being done currently.

Need to make the stadium offset dynamic, and perhaps move it to process_template like the other checksum.

Reply:
This is still hardcoded for now, however it has been added to the patch list in process_template as requested. See https://github.com/pret/pokecrystal/pull/882/commits/a9779bf7fc506da48cfd16edac9dd9a89dbda369

Reply:
Left another comment about this with a solution, so resolving this one.